### PR TITLE
fix(parser): parse decorator with separate grammar

### LINKED ISSUE #191: bug(parser): failed to parse member expression inside decorators.
Found this from a real project.

```
@decorator((baz) => ({
  bar: baz['']
}))
class Foo {}
```

```
  × Expect token
   ╭─[test.ts:1:1]
 1 │ @decorator((baz) => ({
 2 │   bar: baz['']
   ·           ┬
   ·           ╰── Expect `,` here, but found `[`
 3 │ }))
   ╰────
```

### PR DESCRIPTION
**Pull request by YangchenYe323**
Fixes #191

### CODE REVIEW

File: crates/oxc_parser/src/diagnostics.rs [hunk 1]
```diff
@@ -752,3 +752,10 @@ pub struct JSXExpressionsMayNotUseTheCommaOperator(
     #[label("JSX expressions may not use the comma operator. Did you mean to write an array?")]
     pub Span,
 );
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("Leading decorators must be attached to a class declaration")]
+#[diagnostic()]
+pub struct BadLeadingDecorator(
+    #[label("Leading decorators must be attached to a class declaration")] pub Span,
+);
```
Comment (YangchenYe323) [inline crates/oxc_parser/src/diagnostics.rs:761]: Currently we'll falsely accept this code:
```javascript
@a['b']
class A {}
```
Oxc silently eats `@a`, and parses `['b']` as a following ArrayExpression. Personally I'd like to make this part of early error pass. The decorator belongs to the ArrayExpression. And in the early error pass we identify this as illegal. 

But we could also do that directly in the parser.

Comment (Boshen) [reply crates/oxc_parser/src/diagnostics.rs:761]: It's fine to leave it here for now. Only the ones that are on the hot path or have complicated logic are moved to the early pass.

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
I was on travel yesterday so I didn't get the chance to review your code.

This is very well done, it surprises me that this also passes legacy decorators.

**Comment by Boshen** [comment]
You can try and revert 97d01b4 and see if it all passes.

**Review by Boshen — CHANGES_REQUESTED** [review]
Need to fix conflicts.

**Comment by YangchenYe323** [comment]
> You can try and revert [97d01b4](https://github.com/Boshen/oxc/commit/97d01b4ced9b11050291d6efbef3e0c1bf8b5a8e) and see if it all passes.

We're still failing the two related to PrivateIdentifiers inside static blocks.

**Comment by Boshen** [comment]
> > You can try and revert [97d01b4](https://github.com/Boshen/oxc/commit/97d01b4ced9b11050291d6efbef3e0c1bf8b5a8e) and see if it all passes.
> 
> We're still failing the two related to PrivateIdentifiers inside static blocks.

Ok, let's leave it from this PR.

**Comment by Boshen** [comment]
Thank you!

**Pull request merged**